File: 2022/py/day6.py
def solve(s, n):
    for i in range(n, len(s)):
        if len(set(s[i-n : i])) == n:
            return i

with open("input/day6.txt", "r") as f:
    data = f.read().splitlines()
    for line in data:
        print(solve(line, 4))
        print(solve(line, 14))

File: 2022/cpp/day6.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
using namespace std;

int help (string s, int len) {
    int left = 0, right = 0, record[26] = {0};
    while (right - left < len) {
        record[s[right++]-'a']++;
        while (record[s[right-1]-'a'] > 1) record[s[left++]-'a']--;
    }
    return right;
}

int main() {
    ifstream infile;
    infile.open("input/day6.txt", ios::in);
    string cur;
    while (getline(infile, cur)) {
        cout << help(cur, 4) << '\n';
        cout << help(cur, 14) << '\n';
    }
    return 0;
}